### PR TITLE
Add sale_allotment

### PR DESCRIPTION
**Pull request by clonedagain**
This aims to implement partner allotment as discussed in https://github.com/OCA/sale-workflow/issues/206 and https://github.com/odoo/odoo/issues/5371

Forked from the initial implementation by @Openies here: https://www.odoo.com/apps/modules/8.0/line_ship/
- [x] check license for OCA compatibility (OK: AGPL)
- [x] check whether @openies agrees to OCA's CLA

Improvements made (or to make):
- [x] rename the module to `sale_allotment`
- [x] access `procurement_id.sale_line_id.address_allotment_id` directly instead of hijacking `procurement.partner_dest_id`
- [x] adapt to OCA's conventions (README, PEP8...)
- [x] add context to attach quick-created partners as contacts of the customer, with "use the company's address" checked by default
- [x] add the allotment partner to the sale line's treeview (those who don't use other advanced features can't see the field because it's only in the form view)
- [x] make sure all the pickings share the same parameters

### CODE REVIEW

**Comment by clonedagain** [comment]
I think this is ready for review, pending @Openies approval of the CLA.

**Comment by StefanRijnhart** [comment]
Nice module! I confirm @Openis or Anil Makwana cannot be found in the OCA database yet.

**Comment by Openies** [comment]
@Stefan Thanks for the comment.
yes, we have never submitted our module to OCA.

On Wed, Sep 9, 2015 at 12:00 AM, Stefan Rijnhart (Opener) <
notifications@github.com> wrote:

> Nice module! I confirm @Openis or Anil Makwana cannot be found in the OCA
> database yet.
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/OCA/sale-workflow/pull/207#issuecomment-138662048.

## 

Openies
Ingenious Effective Solutions

**Comment by clonedagain** [comment]
@Openies have you had time yet to decide if you agree with OCA's CLA ? it's especially important now that v9 is released.
It'd be perfectly fine by me whether you agree or not: depending on your choice I'd just make a PR to your own branch instead of OCA's.

**Comment by Openies** [comment]
Hello Lionel,

Yes, We are working on that, will revert you soon with CLA.

On Tue, Oct 6, 2015 at 1:22 PM, Lionel Sausin notifications@github.com
wrote:

> @Openies https://github.com/Openies have you had time yet to decide if
> you agree with OCA's CLA ? it's especially important now that v9 is
> released.
> It'd be perfectly fine by me whether you agree or not: depending on your
> choice I'd just make a PR to your own branch instead of OCA's.
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/OCA/sale-workflow/pull/207#issuecomment-145774519.

## 

Openies
Ingenious Effective Solutions

**Comment by Openies** [comment]
Hello Lionel,

We have submitted our Signed OCA CLA.

Openies http://Openies.com.

On Fri, Oct 9, 2015 at 2:18 PM, Openies India openies.com@gmail.com wrote:

> Hello Lionel,
> 
> Yes, We are working on that, will revert you soon with CLA.
> 
> On Tue, Oct 6, 2015 at 1:22 PM, Lionel Sausin notifications@github.com
> wrote:
> 
> > @Openies https://github.com/Openies have you had time yet to decide if
> > you agree with OCA's CLA ? it's especially important now that v9 is
> > released.
> > It'd be perfectly fine by me whether you agree or not: depending on your
> > choice I'd just make a PR to your own branch instead of OCA's.
> > 
> > —
> > Reply to this email directly or view it on GitHub
> > https://github.com/OCA/sale-workflow/pull/207#issuecomment-145774519.
> 
> ## 
> 
> Openies
> Ingenious Effective Solutions

## 

Openies
Ingenious Effective Solutions

**Comment by clonedagain** [comment]
I think I fixed the small coding convention nitpicking, I hope Travis goes green .

**Comment by Openies** [comment]
@clonedagain We have pushed Lineship module to https://github.com/Openies/OdooLineShip/tree/9.0
please have a look.
Thanks!

**Comment by clonedagain** [comment]
@Openies sorry we have no plan to move to v9. Maybe you can propose to migrate/enhance this module directly at OCA's if it gets merged.

**Comment by clonedagain** [comment]
@lbellier can you review this please?

**Comment by lbellier** [comment]
+1

File: sale_allotment/models/stock_move.py [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+        move_rec = self.browse(cr, uid, move_ids, context=context)
+        for move in move_rec:
+            if not partner_dict.get(move.partner_id.id, False):
+                partner_dict.update({move.partner_id.id: [move.id]})
+            else:
+                partner_dict.get(move.partner_id.id).append(move.id)
+        pick_obj = self.pool.get('stock.picking')
+        for key in partner_dict.keys():
+            move = move_rec[0]
+            if partner_dict.keys().index(key) == 0:
```
Comment (StefanRijnhart) [inline sale_allotment/models/stock_move.py]: What is this line supposed to do? It looks like you rely on the order of the keys of the dictionary which is unsafe when using regular dictionaries. You need collections.OrderedDict for that.

File: sale_allotment/models/stock_move.py [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+        partner_dict = {}
+        vals = super(StockMove, self)._picking_assign(
+            cr, uid, move_ids, procurement_group, location_from, location_to,
+            context=context)
+        move_rec = self.browse(cr, uid, move_ids, context=context)
+        for move in move_rec:
+            if not partner_dict.get(move.partner_id.id, False):
+                partner_dict.update({move.partner_id.id: [move.id]})
+            else:
+                partner_dict.get(move.partner_id.id).append(move.id)
```
Comment (StefanRijnhart) [inline sale_allotment/models/stock_move.py]: You could replace these four lines with

```
partner_dict.setdefault(move.partner_id.id, []).append(move.id)
```

File: sale_allotment/models/stock_move.py [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+        pick_obj = self.pool.get('stock.picking')
+        for key in partner_dict.keys():
```
Comment (StefanRijnhart) [inline sale_allotment/models/stock_move.py]: Would be clearer if you renamed 'key' to 'partner_id'

File: sale_allotment/models/stock_move.py [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+                    'move_type': (move.group_id and move.group_id.move_type or
+                                  'direct'),
+                    'partner_id': key or False,
+                    'picking_type_id': (move.picking_type_id and
+                                        move.picking_type_id.id or
+                                        False),
+                }
+                pick = pick_obj.create(cr, uid, values, context=context)
+                self.write(
+                    cr, uid, partner_dict.get(key), {'picking_id': pick},
```
Comment (StefanRijnhart) [inline sale_allotment/models/stock_move.py]: No need to use 'get()'

File: sale_allotment/models/stock_move.py (hunk 1 above)
Comment (clonedagain) [inline sale_allotment/models/stock_move.py]: The standard provides us with a single picking already created.
The way I understand it, it's just a way to assign this picking to one of the groups, and then create new pickings for the other groups.
So the order is not important.

**Comment by lbellier** [comment]
In fact ! This module create pickings without the same header.
It search for user / default parameters only for the first picking, and without for the following others...

**Comment by clonedagain** [comment]
@lbellier @StefanRijnhart I think I addressed your remarks: I've rewritten a part of this module to duplicate the pickings instead of creating new ones and losing the parameters.

**Comment by clonedagain** [comment]
I fixed an api v7/v8 compatibility problem in the code, and added a test to make coverage progress.
Can we merge?

**Comment by lbellier** [comment]
:+1: :+1: !

**Comment by StefanRijnhart** [comment]
Thanks for the update! One last question ;-) Why is there a domain on partners of type 'other'? That would exclude delivery addresses (but contacts might be fine too?).

**Comment by clonedagain** [comment]
@StefanRijnhart, initially we had set the default type and the domain to 'delivery' but users found it error-prone, so we changed both to 'other' instead.
But you made a point, so I've relaxed the domain to "other, delivery or contact" and kept 'contact' as the default.
@lbellier do you agree?

**Comment by StefanRijnhart** [comment]
Thanks. Would you consider squashing the commits per author? It is a bit noisy at the moment in the commits section.

**Comment by lbellier** [comment]
@clonedagain, Yes, i agree.

**Comment by clonedagain** [comment]
@StefanRijnhart: Rebased and squashed, but I've respected @Openies' commits in case they want to merge with their branch.

**Comment by StefanRijnhart** [comment]
Thanks! You did include some pyc files however.

**Comment by clonedagain** [comment]
Oh, the magic of git rebase. I think it's fixed now.

**Comment by StefanRijnhart** [comment]
Thanks! :+1:

**Comment by clonedagain** [comment]
Can this be merged please?

**Pull request merged**